controllers/datafoundationclient.go

```diff
@@ -124,6 +132,29 @@ func (r *dataFoundationClientReconciler) reconcilePhases() (ctrl.Result, error)
 	}
 
 	if !r.offering.DeletionTimestamp.IsZero() {
+		if pvCount, err := r.countOCSVolumes(); err != nil {
+			return ctrl.Result{}, err
+		} else if pvCount > 0 {
+			r.Log.Info("PVs provisioned by ODF storage classes exists, cannot proceed with uninstallation", "Number of PVs", pvCount)
+			return ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
+		}
+
+		storageClassClaimList := &ocsclient.StorageClassClaimList{}
+		if err := r.unrestrictedList(storageClassClaimList); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to list storage class claims: %v", err)
+		}
+		for i := range storageClassClaimList.Items {
+			storageClassClaim := &storageClassClaimList.Items[i]
+			r.Log.Info("Issuing a delete for storageClassClaim %s", storageClassClaim.Name)
+			if err := r.unrestrictedDelete(storageClassClaim); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to issue a delete for storageClassClaim %s: %v", storageClassClaim.Name, err)
+			}
+		}
+
+		r.Log.Info("issuing a delete for StorageClient")
+		if err := r.delete(&r.storageClient); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to issue a delete for Storage Client: %v", err)
+		}
```

Comment:
Do we need this? we are the controller owner of this resource. It will get deleted when the offering CR will get deleted. Am I wrong here?

Reply:
I will try to explain using a scenario
- Let's assume that the ocs-client operator adds a finalizer to a secret for some reason if a storageClient is created.
- This finalizer should be removed by the ocs-client operator if storageClient is in deleting state.

Now, if we delete offering CR that is the owner of storageClient then storageClient will be removed instantly as the owner(offering CR) was removed. At this moment, it is possible that ocs-client operator didn't trigger a reconcile quick enough to remove the finalizer from the secret. This will block deletion of namespace. So according to me, the agent should initiate the deletion of storageClient and after the successful removal of storageClient, we should remove offering CR.

Reply:
We do the same for storageCluster in DF offering

Reply:
What you are describing is a bug in the uninstallation/deletion flow of the client. If the client added a finalizer on a resource it needs to act in one of two consistent ways:
1. remove the finalizer if the client is marked for delete
2. stop uninstallation until the secret is removed. 
Which one is dependent on what it tries to achieve.

Any other behavior will be considered a bug and the client developers will need to fix it.

Reply:
I believe we do the same because we copy-pasted the current behavior form the old deployer where there was no governing offering CR that held an owner ref. We need to evaluate if a change is also needed there.

Reply:
> where there was no governing offering CR that held an owner ref

We had managedOCS CR owning the storageCluster, right?

> eventually consistent model for offering removal or a sync one

Can you elaborate on this?

Reply:
For now, we can follow the behaviour of the deployer. If required, we can evaluate if a change is required for both offerings. I have updated the PR to follow an eventually consistent model for offering removal.